From: Alexander Mikhaylenko <alexander.mikhaylenko@puri.sm>
Date: Tue, 12 Jul 2022 01:53:05 +0400
Subject: window: Maximize resizable transient windows on mobile

---
 gtk/gtkwindow.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/gtk/gtkwindow.c b/gtk/gtkwindow.c
index 66ec112..c32646a 100644
--- a/gtk/gtkwindow.c
+++ b/gtk/gtkwindow.c
@@ -33,6 +33,7 @@
 #include "gtkbuilderprivate.h"
 #include "gtkcheckbutton.h"
 #include "gtkcssshadowvalueprivate.h"
+#include "deprecated/gtkdialog.h"
 #include "gtkdroptargetasync.h"
 #include "gtkeventcontrollerlegacy.h"
 #include "gtkeventcontrollerkey.h"
@@ -109,6 +110,8 @@
 #define GDK_ARRAY_PREALLOC 16
 #include "gdk/gdkarrayimpl.c"
 
+#include "purism-settings.h"
+
 /**
  * GtkWindow:
  *
@@ -4004,6 +4007,7 @@ gtk_window_map (GtkWidget *widget)
   GtkWindow *window = GTK_WINDOW (widget);
   GtkWindowPrivate *priv = gtk_window_get_instance_private (window);
   GtkWidget *child = priv->child;
+  gboolean is_dialog, is_phone;
 
   GTK_WIDGET_CLASS (gtk_window_parent_class)->map (widget);
 
@@ -4019,6 +4023,12 @@ gtk_window_map (GtkWidget *widget)
 
   gtk_window_present_toplevel (window);
 
+  /* Transient windows can be considered pseudo-dialogs. */
+  is_dialog = GTK_IS_DIALOG (widget) || !!gtk_window_get_transient_for (window);
+  is_phone = purism_settings_get_is_phone (purism_settings_get_default ());
+  if (is_dialog && gtk_window_get_resizable (window) && is_phone)
+    gtk_window_maximize (window);
+
   if (priv->minimize_initially)
     gdk_toplevel_minimize (GDK_TOPLEVEL (priv->surface));
 
