From: Alexander Mikhaylenko <alexander.mikhaylenko@puri.sm>
Date: Tue, 12 Jul 2022 01:53:20 +0400
Subject: windowcontrols: Force close buttons for transient windows

---
 gtk/gtkwindowcontrols.c | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/gtk/gtkwindowcontrols.c b/gtk/gtkwindowcontrols.c
index 1f01eb9..f73eef7 100644
--- a/gtk/gtkwindowcontrols.c
+++ b/gtk/gtkwindowcontrols.c
@@ -25,12 +25,14 @@
 #include "gtkboxlayout.h"
 #include "gtkbutton.h"
 #include "gtkenums.h"
+#include "deprecated/gtkdialog.h"
 #include "gtkicontheme.h"
 #include "gtkimage.h"
 #include <glib/gi18n-lib.h>
 #include "gtkprivate.h"
 #include "gtktypebuiltins.h"
 #include "gtkwindowprivate.h"
+#include "purism-settings.h"
 
 #ifdef GDK_WINDOWING_MACOS
 #include "gtkwindowbuttonsquartzprivate.h"
@@ -134,9 +136,17 @@ get_layout (GtkWindowControls *self)
   if (self->decoration_layout)
     layout_desc = g_strdup (self->decoration_layout);
   else
-    g_object_get (gtk_widget_get_settings (widget),
-                  "gtk-decoration-layout", &layout_desc,
-                  NULL);
+    {
+      gboolean is_phone = purism_settings_get_is_phone (purism_settings_get_default ());
+      gboolean is_mobile_dialog = GTK_IS_DIALOG (root) || !!gtk_window_get_transient_for (GTK_WINDOW (root));
+
+      if (is_phone && is_mobile_dialog)
+        layout_desc = g_strdup (":close");
+      else
+        g_object_get (gtk_widget_get_settings (widget),
+                      "gtk-decoration-layout", &layout_desc,
+                      NULL);
+    }
 
   if (layout_desc == NULL || layout_desc[0] == '\0')
     {
@@ -644,6 +654,12 @@ gtk_window_controls_init (GtkWindowControls *self)
   gtk_widget_add_css_class (GTK_WIDGET (self), "start");
 
   gtk_widget_set_can_focus (GTK_WIDGET (self), FALSE);
+
+  g_signal_connect_object (purism_settings_get_default (),
+                           "notify::is-phone",
+                           G_CALLBACK (update_window_buttons),
+                           self,
+                           G_CONNECT_SWAPPED);
 }
 
 /**
